dm/pkg/checker/conn_checker.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checker
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/pingcap/tiflow/dm/dm/config"
+	"github.com/pingcap/tiflow/dm/pkg/conn"
+	tcontext "github.com/pingcap/tiflow/dm/pkg/context"
+	"github.com/pingcap/tiflow/dm/pkg/log"
+)
+
+type connAmountChecker struct {
+	targetDB *conn.BaseDB
+	stCfgs   []*config.SubTaskConfig
+
+	getConfigConn func(stCfgs []*config.SubTaskConfig) int
+	workerName    string
+}
+
+func newConnAmountChecker(targetDB *conn.BaseDB, stCfgs []*config.SubTaskConfig, fn func(stCfgs []*config.SubTaskConfig) int, workerName string) connAmountChecker {
+	return connAmountChecker{
+		targetDB:      targetDB,
+		stCfgs:        stCfgs,
+		getConfigConn: fn,
+		workerName:    workerName,
+	}
+}
+
+func (c *connAmountChecker) check(ctx context.Context, checkerName string) *Result {
+	result := &Result{
+		Name:  checkerName,
+		Desc:  "check if user-specified amount of connection exceeds downstream's maximum",
+		State: StateFailure,
+	}
+	baseConn, err := c.targetDB.GetBaseConn(ctx)
+	if err != nil {
+		markCheckError(result, err)
+		return result
+	}
+	var rows *sql.Rows
+	rows, err = baseConn.QuerySQL(tcontext.NewContext(ctx, log.L()), "SHOW VARIABLES LIKE '%max_connections%'")
```

Comment:
You can reuse func `GetMaxConnections`.

Reply:
Yes. But `GetMaxConnections` has no retry strategy. Users have to retry by themselves if connection errors occur (#5015).

Actually, there are a bunch of different pkgs used for querying SQL, varying from implementation and retry strategy. `BaseConn` is good to have decent retry strategy as well as encapsulation. Should we refactor and integrate them as a whole package?

Reply:
lgtm

Reply:
Currently, @lance6716 is planning to refactor the syncer. Idk if this could be part of the plan cuz it seems like a big one to do.